Lobby/GUI work

SJuliez opened the pull request:
OK I'm finishing this now (tentatively) to at least start the process of integrating it.

I have no clue why the Messages call in the Boardvalidator fails.

Main Changes (omitting small bug fixes in the tooltip and elsewhere):
- Unifies the tooltip in the lobby and game (while still enabling different views, e.g. the pilot info can be omitted as is done in the lobby mek table) and moves the code out of Entity; the tooltip now shows a different symbol for capital armor. I've introduced many unicode symbols, these require the Java "Dialog" logic font, I hope they work on other systems besides Win10 as well. 
- Adds a general gui scaling float value (neutral value = 1) that is accessible in the Client Settings and, unless a modal dialog is up, with Ctrl+Numpad+ and Ctrl+Numpad- . This currently scales mainly the lobby and some dialogs in the lobby that have been revamped. The scaling can be applied by calling UIUtil.adjustDialog on any dialog, but the result depends on how complicated the dialog is and its probably better to adjust manually. I've done this with the host game dialog and the result is I guess ok.
- Completely revamps the info in the lobby mektable, obviously. I've tried using colors that are desaturated in the hope this will make the colors readable for everyone. Also, the main colors come from methods that return different colors based on whether the current UI theme is light or dark.
- Allows sorting the mek table by various algorithms
- Generally shows more info directly in the table such as velocity, height, deployment, partial repairs, C3 info. I've tried to hide stuff thats unnecessary such as height in space and so on to make it all kinda useful
- The lobby now shows if a unit has a problem such as instant death upon deployment or an unconnected C3 computer; the tooltip gives details on the problem (unit cannot survive in vacuum)
- Makes the compact mode more useful; it now shows alomst the same info as the full mode.
- Revamps the popup menu; it now allows some more unit settings such as C3 configuration and most operations work on multiple units
- Adds toggles to hide unofficial and legacy game options; also adds a ToggleButton that has a checkmark or cross as the normal Java togglebuttons dont show their status really well
- Adds searchlights automatically to all units in circumstances when they're needed
- Revamps the player config and planetary settings dialog
- Revamps the player table; it now allows multi-selection and direct setting of team and deployment via popup, also shows local bots with a symbol
- Makes some dialogs rememeber their size and position such as the Unit selector and force generator
- Adds Ctrl-S quicksave and Ctrl-L quickload. I think this is only useful for development. Ctrl-L works only from the main menu I think.
- The mektable generally remembers column sizes
- removes some buttons from the main interface (such as Delete All) and adds a confirm dialog to delete and various other things that shouldnt be insta-kill
- Minefields now actually require the TacOps minefields game option to be active
- Corrects conventional fighter weight classes, use the same as ASF (50t = medium)
- Corrects legal deployment with double blind
- Obviously revamps board selection to graphical dragndrop. It can deal with server-side boards (although in this case it can't actually display the real board, just an empty placeholder) and invalid boards. 
- Instead of choosing a board at random, the surprise function will now choose a board from a list of boards that can be assigned by selecting the desired boards and dragging them to the preview.
- Board setups can now be saved. I've created some examples from the boards we have
- Adds a smaller minimap zoom that I neede for those huge maps and a better space minimap with stars...
- Adds a team overview pane. The table can be detached to a window. It shows the summarized strength of the teams and when one team is selected in the table it shows the relative values (87% of ...)
- Adds ammo to the unit tooltips
- Adds forces to MM and a force tree view to the lobby. Forces currently have no ingame effect. The game has a Forces object that controls the individual Force objects. The structure is similar to MekHQ's forces although I believe MekHQ has no owner field. Forces can be saved to MUL files and load from there correctly. However it only saves the full force every unit is in, not the forces themselves. When loading a MUL, the force structure is reconstructed from these individual force entries. Therefore, loading from a MUL will always create new forces, never integrate into existing ones. The forces from MekHQ can be transported to MM by using the String representation as Ive done in the MekHQ PR. The MekHQ PR is not required for this PR; it can be pulled later when MM actually works.
- Units in forces and subforces can be moved up and down by Ctrl-Up and Ctrol-Down; no multi-selection.
- The mek table and force table allow Ctrl-C to copy selected units (forces are ignored). It exports the unit, tonnage, pilot and a few values and pastes mostly well into Excel, although some units have secondary names that dont paste well such as "(3025)" which arrives as -3025 in Excel. Ive found no solution for this that also allows pasting into MM well. 
- Obviously, the mektable and force table also allow Ctrl-V. So it's easy to copy-paste units in MM. 
- Also, makes the keybinds overlay disappear in the map preview


Resolves #2384 
Resolves #2385 
Resolves #2340 
Resolves #2327 
Resolves #2322 
Resolves #2306 
Resolves #2277 
Resolves #2428 
Resolves #1996
Resolves #448 
Resolves #2560 
Resolves #765 
Resolves #2657 
Resolves #2374

I believe this also resolves #322. It does not enable a GM player to see all units during blind drop, but the team overview allows a cost/tonnage/BV comparison through all blind drop options.

![image](https://user-images.githubusercontent.com/17069663/99563465-93d04f80-29c9-11eb-98d8-4a6568c3491f.png)

HammerGS commented:
This looks great. One first impression from the screen shot, I think the unit ID would better after the unit name.

SJuliez commented:
Thanks. Note that on the left side theres a toggle button for the unit ID,  so it's easily turned off and on (took this from the CasualJoker vids) and the unit name can be very (very) long. I can try this though.

HammerGS commented:
Curious would this one fit into the last of UI stuff - https://github.com/MegaMek/megamek/issues/2183

sixlettervariables commented:
CI is failing because the source compatibility level is 8 and you have some >8 Java-isms.

HammerGS commented:
Wanted to have a look. But looks like it has conflicts.

```
Auto-merging megamek/src/megamek/server/Server.java
Auto-merging megamek/src/megamek/common/Game.java
Auto-merging megamek/src/megamek/common/Entity.java
Auto-merging megamek/src/megamek/client/ui/swing/gameConnectionDialogs/HostDialog.java
Auto-merging megamek/src/megamek/client/ui/swing/gameConnectionDialogs/ConnectDialog.java
Auto-merging megamek/src/megamek/client/ui/swing/dialog/MegaMekUnitSelectorDialog.java
Auto-merging megamek/src/megamek/client/ui/swing/dialog/AbstractUnitSelectorDialog.java
Auto-merging megamek/src/megamek/client/ui/swing/boardview/BoardView1.java
Auto-merging megamek/src/megamek/client/ui/swing/MegaMekGUI.java
Removing megamek/src/megamek/client/ui/swing/MechGroupView.java
Auto-merging megamek/src/megamek/client/ui/swing/GUIPreferences.java
Auto-merging megamek/src/megamek/client/ui/swing/FiringDisplay.java
Auto-merging megamek/src/megamek/client/ui/swing/EquipChoicePanel.java
Auto-merging megamek/src/megamek/client/ui/swing/CommonSettingsDialog.java
Auto-merging megamek/src/megamek/client/ui/swing/ClientGUI.java
CONFLICT (modify/delete): megamek/src/megamek/client/ui/swing/ChatLounge.java deleted in 4f5175175a282f66f2e6f36300bcf038899dfd36 and modified in HEAD. Version HEAD of megamek/src/megamek/client/ui/swing/ChatLounge.java left in tree.
Auto-merging megamek/src/megamek/client/bot/princess/Princess.java
Auto-merging megamek/i18n/megamek/client/messages.properties
Automatic merge failed; fix conflicts and then commit the result.
```

SJuliez commented:
Hopefully better now. Not many changes, I struggled with the mektable popup and most of the changes there are not so visible. It'll only show certain load options when a unit type is present, like fighter squadrons. Squadrons require Capital Fighters (which Im not sure is really what the rules meant to say...). Something visible is the player config dialog. To get the most out of it, activate minefields (this option now really toggles minefields in the lobby) and double blind with exlcusive deployment and open it for a bot client.

HammerGS commented:
Thanks, really looking forward to see this done. More I use it the harder it is to go back to the old lobby.

HammerGS commented:
Loving the improvement but think I got a bug.  Running from repo, add a unit to the lobby. Right click and select view. Nothing happens.  Log has the following error.

```
Action! javax.swing.JButton[,6,22,380x33,alignmentX=0.0,alignmentY=0.5,border=com.formdev.flatlaf.ui.FlatButtonBorder@1539f6af,flags=288,maximumSize=,minimumSize=,preferredSize=,defaultIcon=,disabledIcon=,disabledSelectedIcon=,margin=javax.swing.plaf.InsetsUIResource[top=2,left=14,bottom=2,right=14],paintBorder=true,paintFocus=true,pressedIcon=,rolloverEnabled=true,rolloverIcon=,rolloverSelectedIcon=,selectedIcon=,text=Add A Combat Unit...,defaultCapable=true]
Action! javax.swing.JButton[,183,5,137x29,alignmentX=0.0,alignmentY=0.5,border=com.formdev.flatlaf.ui.FlatButtonBorder@1539f6af,flags=288,maximumSize=,minimumSize=,preferredSize=,defaultIcon=,disabledIcon=,disabledSelectedIcon=,margin=javax.swing.plaf.InsetsUIResource[top=2,left=14,bottom=2,right=14],paintBorder=true,paintFocus=true,pressedIcon=,rolloverEnabled=true,rolloverIcon=,rolloverSelectedIcon=,selectedIcon=,text=Game Options...,defaultCapable=true]
[Fatal Error] log4j.xml:2:10: DOCTYPE is disallowed when the feature "http://apache.org/xml/features/disallow-doctype-decl" set to true.
```

SJuliez commented:
Thanks! found it ... I wanted to check if exactly one unit is selected and instead checked if zero units are selected.

HammerGS commented:
Huge improvement on the planet dialog. One cool thing would be to pull the image used from MekHQ and the planet they are on if available :) 

Also showed a preview on the Casual Jokers Stream lots of kudos.

HammerGS commented:
That update to the map selector is AMAZING?  Is there any way to add the name to them? Under the map maybe?

SJuliez commented:
Sure, I'll take a look. The names can be awfully long though, so some of them will be cut off.

HammerGS commented:
Finding something odd with the map selector. It does consistently select a map. Nothing in the log but will try and duplicate it. It's also not clear to drag and drop the map, and double click doesn't seem to do anything.

HammerGS commented:
Another suggestion for the map selector. Since we have the search, could we have an "all maps" added to the map size.

SJuliez commented:
> It does consistently select a map.
Im not sure what you mean, but this was not finished by any means, I just pushed it because I switched branches so any odd behaviour is possible. Next version wil be much better.
> All maps
I'll have to see how the map list behaves because the map sizes have such a huge range. Also, this will be problematic with how to handle multiple selected boards (you can drag multiple boards in my version; I'll upload it shortly, but this requires that boards are the same size. Not sure how to handle this). Maybe I can allow any board size as long as the map is 1 board.
> Merge
Yes, I'll have to do that, but Im not looking forward to it, because merging when files were moved (I moved the ChatLounge to its own package) is not nice (or I just don't know how to do that correctly which is also possible).

HammerGS commented:
I haven't been able to duplicate the map issue I was having. I think the names look great and we might test need to shorten some of the map names.

I have a Lobby question for you. This weekend we had a big game 14 players, each with 2 mechs but set up on 2 teams. Could we have a way to total up the BV by team.

![image](https://user-images.githubusercontent.com/13752452/104139793-0840c280-536b-11eb-86dc-f6c06deba4c9.png)

In the picture you can see the BV totals for Team 1 are split between players.  Maybe something like in red boxes have

4534 BV (Team total 14324 BV)

Also this shaping up to be an amazing improvement to the lobby.

NickAragua commented on megamek/src/megamek/client/Client.java:
```diff
@@ -1757,4 +1884,18 @@ public void setCurrentHex(IHex hex) {
+        return bots.containsKey(player.getName());
+    }
+    
+    /** Returns true when the player is a bot added/controlled by this client. */
```
Comment doesn't match method behavior. Also, do we really want to be using exceptions as flow control?

SJuliez replied:
I'll change it to simply return null when the given player isn't actually a bot. Saves space. I don't mind but for the sake of clarity: When a method can't return a value that is correct, then why not use an exception? It just means it'll fail clearly when the caller doesn't do the appropriate checks before. In the same case, returning null will likely lead to an NPE, just at another place. I wouldn't know what else to return.

NickAragua replied:
Looking at it again, in this particular case, I don't think I have a problem with the exception, just the mismatched comment. Maybe include an update to the comment that indicates this throws an exception.

NickAragua commented on megamek/src/megamek/client/ui/swing/ClientDialog.java:
```diff
@@ -51,6 +53,12 @@ public ClientDialog(JFrame owner, String title, boolean modal) {
         this.owner = owner;
     }
+    
+    public ClientDialog(JFrame owner, String title, boolean modal, boolean scale) {
```
Is it possible to get some explanation on why somebody would use one of these constructors over the others?

NickAragua commented on megamek/src/megamek/client/ui/swing/ClientGUI.java:
```diff
@@ -709,6 +704,9 @@ public void actionPerformed(ActionEvent event) {
                 saveGame();
                 break;
+            case FILE_GAME_QSAVE:
+                quickSaveGame();
```
Sneaky a little extra feature in there. Nice.

NickAragua commented on megamek/src/megamek/client/ui/swing/ClientGUI.java:
```diff
@@ -1006,12 +1020,6 @@ void die() {
         }
         client.die();
 
-        // TODO Is there a better solution?
```
I guess there was a better solution. What happens now?

SJuliez replied:
From the former comment:
// This is required because the ChatLounge adds the listener to the
// MechSummaryCache that must be removed explicitly.
I think the call to `((Distractable) component).removeAllListeners();` a few lines above already removes that listener.

NickAragua commented on megamek/src/megamek/client/ui/swing/CommonMenuBar.java:
```diff
@@ -243,6 +247,16 @@ public CommonMenuBar() {
+        fileGameQSave.setActionCommand(ClientGUI.FILE_GAME_QSAVE);
+        fileGameQSave.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_S, InputEvent.CTRL_DOWN_MASK));
+        submenu.add(fileGameQSave);
+        fileGameQLoad = new JMenuItem(Messages.getString("CommonMenuBar.fileGameQuickSave")); //$NON-NLS-1$
```
This should probably be QuickLoad instead.

SJuliez replied:
Yes.

NickAragua commented on megamek/src/megamek/client/ui/swing/CommonMenuBar.java:
```diff
@@ -937,19 +964,19 @@ private synchronized void manageMenu() {
         if (hasUnitList) {
             fileUnitsOpen.setEnabled(phase == IGame.Phase.PHASE_LOUNGE);
-            fileUnitsClear.setEnabled(phase == IGame.Phase.PHASE_LOUNGE);
           //  fileUnitsSave.setEnabled(phase == IGame.Phase.PHASE_LOUNGE);
```
Maybe get rid of the commented out code here and on line 980? Or address it somehow otherwise.

SJuliez replied:
I think I'll remove these menu items. They are for the lobby only anyway and saving unit lists depends on blind drop options which the menu bar doesnt adress and loading units lists depends on the selected player which the menu bar also doesnt adress. There's big buttons for loading/saving.

NickAragua commented on megamek/src/megamek/client/ui/swing/GameOptionsDialog.java:
```diff
@@ -30,24 +30,9 @@
 import java.awt.event.WindowAdapter;
```
Think this one needs the megamek license header.

NickAragua commented on megamek/src/megamek/client/ui/swing/PlanetaryConditionsDialog.java:
```diff
@@ -20,544 +20,618 @@
+        //        if (temp >= 30 && requiresLowTemp(weather)) {
+        //            tempValid = false;
+        //            wthrValid = false;
+        //            tempTip.append("The Temperature cannot be 30 °C or more in snowy weather.<BR>");
```
I'm still amused at the idea of water-based snow in 30C weather.

NickAragua commented on megamek/src/megamek/client/ui/swing/PlanetaryConditionsDialog.java:
```diff
@@ -20,544 +20,618 @@
-                                    .getString("PlanetaryConditionsDialog.NumberFormatError"),
-                            JOptionPane.ERROR_MESSAGE);
-            return false;
+        if ((atmo == ATMO_TRACE) && (wind == WI_LIGHT_GALE)) {
```
How about vacuum and wind?

SJuliez replied:
Vacuum disables all weather settings, so it doesnt need to be checked there. But it was not actually resetting the weather. Will be corrected.

NickAragua commented on megamek/src/megamek/client/ui/swing/PlanetaryConditionsDialog.java:
```diff
@@ -20,544 +20,618 @@
+    
+    /** 
+     * Updates the enabled state of some fields based on the atmosphere setting. 
+     * Also, for some weather settings,   
```
Also, for some weather settings what?

NickAragua commented on megamek/src/megamek/client/ui/swing/RandomArmyDialog.java:
```diff
@@ -536,20 +537,25 @@ public RandomArmyDialog(ClientGUI cl) {
         // construct the main dialog
         setLayout(new BorderLayout());
-        this.setPreferredSize(new Dimension(800,500));
+//        this.setPreferredSize(new Dimension(800,500));
```
Why's this commented out, along with the pack() on line 550?

SJuliez replied:
The whole dialog now remembers its position and size, therefore the setpreferredsize and pack are unnecessary. I'll remove the lines.

NickAragua commented on megamek/src/megamek/client/ui/swing/RandomArmyDialog.java:
```diff
@@ -784,34 +790,51 @@ public void actionPerformed(ActionEvent ev) {
-    public void windowIconified(WindowEvent arg0) {
-        //ignored
-    }
+//    public void windowActivated(WindowEvent arg0) {
```
If we don't need these, how about deleting them?

SJuliez replied:
done.

NickAragua commented on megamek/src/megamek/client/ui/swing/boardview/EntitySprite.java:
```diff
@@ -718,496 +688,6 @@ public Coords getPosition() {
         return entity.getPosition();
     }
     
-    private StringBuffer tooltipString;
```
Where'd all this stuff go?

SJuliez replied:
To UnitTooltip

NickAragua commented on megamek/src/megamek/client/ui/swing/dialog/AbstractUnitSelectorDialog.java:
```diff
@@ -187,7 +187,10 @@ private void setUserPreferences() {
 
         tableUnits.invalidate(); // force re-layout of window
-        pack();
+//        pack();
```
Commented out pack?

SJuliez replied:
Same as the RandomArmyDialog, this dialog now remembers size&position and the pack had to go to make it work.

NickAragua reviewed (changes requested):
Got through about half the files. Didn't review the data files, since I wouldn't know what's correct and what isn't anyway. Will take a look at the rest of the files over the next few days.

NickAragua commented on megamek/src/megamek/client/ui/swing/lobby/LobbyActions.java:
```diff
@@ -0,0 +1,1312 @@
+     * fighters.
+     */
+    private IPlayer createSquadronOwner(Collection<Entity> entities) {
+        if (entities.stream().anyMatch(e -> e.getOwner().equals(localPlayer()))) {
```
Gotta ask, is there a reason we're guessing who owns the squadron that a given entity is in? Do we allow entities from multiple players to form squadrons or something?

SJuliez replied:
Yes, loading/squadron/C3/forces are allowed within teams. I didn't want to add an owner dropdown or something for a fighter squadron. Most of the time there'll only be one owner anyway.

NickAragua commented on megamek/src/megamek/client/ui/swing/lobby/LobbyMekPopupActions.java:
```diff
@@ -0,0 +1,328 @@
+        switch (command) {
+            
+            // Single entity commands
+        case "CONFIGURE":
```
Well, I see where all the magic strings come from.

NickAragua commented on megamek/src/megamek/server/WeatherProcessor.java:
```diff
@@ -113,7 +113,6 @@ private void resolveWeather() {
             }
         }
         if(((conditions.getWeather() == PlanetaryConditions.WE_HEAVY_SNOW)
-                || (conditions.getWeather() == PlanetaryConditions.WE_BLIZZARD)
```
Where'd blizzard go?

SJuliez replied:
Global Warming? I just couldn't find it in the rules anymore.

NickAragua commented on megamek/src/megamek/server/ServerHelper.java:
```diff
@@ -449,4 +441,510 @@ public static void checkAndApplyMagmaCrust(IHex hex, int elevation, Entity entit
+     * mapSettings. The path names are minus the '.board' extension and relative to
+     * the boards data directory.
+     */
+    static ArrayList<String> scanForBoards(MapSettings mapSettings) {
```
In general, I'm kind of concerned that this file will become another disorganized mess like Server.java. I think it'd be better to organize the various methods into different ServerHelper classes, e.g. the board scanning/loading related methods into ServerBoardHelper, etc.

SJuliez replied:
Yes I hijacked this as my Eclipse continually goes on 10s hibernations when I edit anything in Server. I'll see if I can divide some things out.

NickAragua commented on megamek/src/megamek/server/Server.java:
```diff
@@ -29964,13 +29868,38 @@ public boolean accept(Entity entity) {
+                    }
+                    topLevel = false;
+                }
+                entity.setForceString("");
```
We no longer have a need for this once we load the entity's forces?

NickAragua commented on megamek/src/megamek/server/Server.java:
```diff
@@ -30435,11 +30467,29 @@ private void receiveEntityAmmoChange(Packet c, int connIndex) {
 
-            // Only allow players to delete their *own* entities.
-            if ((entity != null) && (entity.getOwner() == getPlayer(connIndex))) {
+            // Players can delete units of their teammates
```
Sounds like a great prank.

NickAragua commented on megamek/src/megamek/common/util/CollectionUtil.java:
```diff
@@ -0,0 +1,52 @@
+     * Throws a NoSuchElement exception if it is empty.
+     */
+    public static final <T> T randomElement(Collection<T> collection) {
+        return collection.stream().findFirst().get();
```
Is this really random?

SJuliez replied:
Renamed.

NickAragua commented on megamek/src/megamek/common/Terrains.java:
```diff
@@ -211,38 +211,16 @@ public static String getEditorTooltip(int type) {
-            } else if (level == 3) {
-                return "Heavy building";
-            } else if (level == 4) {
-                return "Hardened Building";
```
I assume that this descriptive building text is contained somewhere else?

SJuliez replied:
No, this was not a good change. I'll revert it.

NickAragua commented on megamek/src/megamek/common/PlanetaryConditions.java:
```diff
@@ -48,13 +48,14 @@
+    public static final int WE_LIGHT_HAIL       = 12;// NYI
+    public static final int WE_HEAVY_HAIL       = 13;// NYI
+    public static final int WE_LIGHTNING_STORM  = 14;// NYI
+    //  public static final int WE_BLIZZARD         = 11; does not exist anymore
```
Guess I know what happened to blizzard

NickAragua commented on megamek/src/megamek/common/OffBoardDirection.java:
```diff
@@ -78,4 +78,30 @@ public static OffBoardDirection getOpposite(OffBoardDirection value) {
+     * position as defined in IStartingPositions.
+     */
+    public static OffBoardDirection translateStartPosition(int startPos) {
+        if (startPos > 10) {
```
I'm having a little trouble understanding the logic in this method.

Windchild292 replied:
I think it might mean modulo ten?

SJuliez replied:
The deep deployment zones use the numbers of the normal deployment zones +10

NickAragua commented on megamek/src/megamek/common/EntityWeightClass.java:
```diff
@@ -252,8 +250,6 @@ public static int getWeightClass(double tonnage, Entity en) {
                     break;
                 }
             }
-        } else if (en instanceof ConvFighter) {
```
In the other method, you added conventional fighters to the aero else if, here, they're just removed?

SJuliez replied:
ConvFighter is instanceof Aero so the Aero part works for ConvFighters in the second method.

NickAragua reviewed (changes requested):
Round 2. 98/132 files reviewed.

NickAragua commented on megamek/src/megamek/client/ui/swing/tooltip/UnitToolTip.java:
```diff
@@ -0,0 +1,870 @@
+    }
+    
+    /** Returns a list of units loaded onto this unit. */
+    private static StringBuilder forceEntry(Entity entity, IPlayer localPlayer) {
```
Comment doesn't match method name or behavior.

NickAragua commented on megamek/src/megamek/client/ui/swing/util/UIUtil.java:
```diff
@@ -0,0 +1,735 @@
+        return "<HTML>" + UIUtil.guiScaledFontHTML() + Messages.getString(str) + "</FONT></HTML>";
+    }
+    
+    public static void adjustDialog(Container contentPane) {
```
What's this do?

SJuliez replied:
This applies the current gui scale to a dialog by setting fonts and adjusting EmptyBorders. This works kinda well with simple dialogs (the host game dialog, for example).

NickAragua reviewed (changes requested):
All right, finished. Some questions and relatively minor change requests.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+        grpMap.add(butLowAtmoMap);
+        grpMap.add(butHighAtmoMap);
+        grpMap.add(butSpaceMap);
+//        panMap.add(panMapType);
```
Its added a few lines below. Just leftovers from moving stuff around. Will delete those lines.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+        panSettings.setAlignmentX(JPanel.CENTER_ALIGNMENT);
+        panSettings.add(butConditions);
+        panSettings.add(butRandomMap);
+//        panMap.add(panSettings);
```
As above.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+    private void removeBot() {
+        Client c = getSelectedClient();
+        if (!client().bots.containsValue(c)) {
+//        if ((c == null) || c.equals(clientgui.getClient())) {
```
Leftovers. Will remove.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+                clientgui.getClient().sendMapDimensions(mapSettings);
+            }
+        } catch (NumberFormatException e) {
+            // no number, no new map width
```
you mean like `boolean tryParse(String source, Object resultValue) {}`? It would help if Java could handle return tupels. (Or anything else C# added in the last few years :-) )

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+                e.consume();
+                lobbyActions.customizeMech(selEntities.get(0));
+                
+            } else if (code == KeyEvent.VK_UP && e.getModifiersEx() == InputEvent.CTRL_DOWN_MASK) {
```
No. I could add this to the popup with the key shortcut so people will see it.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+            }
+            
+            // Found or created an icon; finish the panel
+            setText("");
```
When an icon isn't there, the super renderer result is returned a few lines above. This will set the text, so it has to be reset when an icon is found and used (otherwise, the text of a first list entry will stay present for all subsequently rendered list entries, sorry if that overexplains).

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/ChatLounge.java:
```diff
@@ -0,0 +1,3628 @@
+     * Returns a list of the selected entities in the Mek table. 
+     * The list may be empty but not null. 
+     */
+    private ArrayList<Entity> getSelectedEntities() {
```
OK done, though I wonder at this practice (I also found it at some Java Practices website). Why stop at List? If I change the method to actually use a Set (which I've done multiple times while doing the lobby stuff) I still have to change the return value and possibly the callers. Isnt it the fault of the caller to not use a Collection whenever possible? If they do, I can return an ArrayList or a HashSet, no problem. Shouldnt the actual practice be to always receive a Collection?

Windchild292 replied:
The reason I code using List/Set instead of Collection is that both give you more information and functionality while still being generic.

If you see a set, you know the values within are unique and that contains is a fast method to use.
If you see a list, you know the values within are ordered, sortable, accessible and insertable by index, and searchable by element.

The individual implementations within those two are less important, and can thus be genericized out, but just having a collection (especially with list) removes a huge amount of functionality and information for the programmer.

So... I'd use collection only if you don't and cannot think of a future reason why to use any list-based functionality and don't care about the duplication state (which... would just be easier to use and return List still to future proof it).

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/LobbyActions.java:
```diff
@@ -0,0 +1,1312 @@
+        if (!validateUpdate(entities)) {
+            return;
+        }
+        Entity entity = CollectionUtil.randomElement(entities);
```
Here and at other places, I wanted to receive Collections instead of lists (as discussed above), especially as the order of the selected entities shouldnt play any role. When the collection has only 1 element its still bothersome to get that element, therefore I used that method. Sometimes I used it with multiple entities to find some, e.g. to find one owner. I think I'll rename it to anyOneElement or something like that as it really has nothing to do with random.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/LobbyActions.java:
```diff
@@ -0,0 +1,1312 @@
+    }
+    
+    public void load(Collection<Entity> selEntities, String info) {
+        StringTokenizer stLoad = new StringTokenizer(info, ":");
```
Yup, will remove from ChatLounge

Windchild292 commented on megamek/data/images/temp/The images in this directory can be safely deleted between games.txt:
```diff
@@ -0,0 +1 @@
+The images in this directory ar temporary images used for the pilot tooltips. They can be safely deleted while not in a MegaMek game.
```
```suggestion
The images in this directory are temporary images used for the pilot tooltips. They can be safely deleted while not in a MegaMek game.
```

Windchild292 replied:
Still holds true

Windchild292 commented on megamek/src/megamek/client/ui/swing/ReportDisplay.java:
```diff
@@ -223,10 +228,11 @@ public void setReportTab(int round, String roundText, String phaseText) {
+        int size = UIUtil.scaleForGUI(UIUtil.FONT_SCALE1);
         ((HTMLEditorKit) pane.getEditorKit()).getStyleSheet().addRule(
-               "pre { font-family: " + font.getFamily() + "; font-size: 12pt; font-style:normal;}");
+                "pre { font-family: " + font.getFamily() + "; font-size: "+size+"pt; font-style:normal;}");
```
```suggestion
                "pre { font-family: " + font.getFamily() + "; font-size: " + size + "pt; font-style:normal;}");
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/MovementDisplay.java:
```diff
@@ -2116,12 +2116,12 @@ private void updateRACButton() {
-            return;
-        }
+        if (ce != null) {
+        boolean isNight = clientgui.getClient().getGame().getPlanetaryConditions().isSearchlightEffective();
         setSearchlightEnabled(
```
Indent

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -206,8 +214,8 @@ BoardSelectionDialog.UpdateMapSize.message=Please hit the Update Map Size button
-BoardSelectionDialog.ViewGameBoard=View Game Board*
-BoardSelectionDialog.ViewGameBoardTooltip=<html>Shows a preview of the selected map.<br>  *Note: for generated maps the preview is only a sample<br> and will not be the same as the actual map used.</html>
+BoardSelectionDialog.ViewGameBoard=Preview Game Map... *
+BoardSelectionDialog.ViewGameBoardTooltip=Shows a preview of the game's map.<br>  * Note: For generated maps and surprise maps the preview is only a sample<br> and will not be the same as the actual map used.
```
Recommendation: Swap the side where the BR is, to avoid having 
```
Hello
 World
```

SJuliez replied:
The "* Note" is supposed to be indented.

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -344,58 +352,49 @@ BoardView1.ROLLED=ROLLED
-BoardView1.Tooltip.Adv1=<BR><FONT SIZE=-2 COLOR=#555555>&nbsp;&nbsp;1 Adv.;</FONT>
-BoardView1.Tooltip.PilotPortraitW=</TR><TR><TD VALIGN=\"top\"><IMG WIDTH=60 HEIGHT={1} SRC=\"file:{0}\"></TD>
-BoardView1.Tooltip.PilotPortraitH=</TR><TR><TD VALIGN=\"top\"><IMG WIDTH={1} HEIGHT=60 SRC=\"file:{0}\"></TD>
+BoardView1.Tooltip.AeroVelAlt=<I>Velocity: {0}, Altitude: {1}</I>
+BoardView1.Tooltip.AeroStartingVelAlt=<I>Starting Velocity: {0}, Altitude: {1}</I>
 BoardView1.Tooltip.ArmorInternals=Armor {0}; Internal {1}
+BoardView1.Tooltip.ArmorCapital=<I>Capital Armor</I>
```
Italics are terrible for accessibility, and should **never** be used. Use em or b instead.

Windchild292 replied:
This holds for the whole file, as I see a number of other cases where this occurs.

SJuliez replied:
This has been like that before, lets maybe discuss outside. Easily changed.

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -550,22 +567,55 @@ ChatLounge.PartialRepairs=Partial Repairs
+ChatLounge.teamOverview.unitSum1=Vehicles:
+ChatLounge.teamOverview.unitSum2=Aero:
+ChatLounge.teamOverview.unitSum3=Infantry:
+ChatLounge.teamOverview.unitSum4=Protomeks:
```
```suggestion
ChatLounge.teamOverview.unitSum4=ProtoMeks:
```

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -1266,6 +1325,7 @@ GameOptionsDialog.Save=Save
 GameOptionsDialog.Search=Search
 GameOptionsDialog.SearchToolTip=Search for options that contain the entered text
 GameOptionsDialog.Load=Load
+GameOptionsDialog.HideWarning=Warning: All hidden options will also be turned off. If you reenable these options, you will have to reenable any of the affected options manually. Are you sure?
```
```suggestion
GameOptionsDialog.HideWarning=Warning: All hidden options will also be turned off. If you re-enable these options, you will have to re-enable any of the affected options manually. Are you sure?
```

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -2090,23 +2150,37 @@ PlanetaryConditionsDialog.labAtmosphere=Atmospheric Pressure
+PlanetaryConditionsDialog.labGrav=Gravity (g)
 PlanetaryConditionsDialog.labLight=Light
-PlanetaryConditionsDialog.labTemp=Temperature
+PlanetaryConditionsDialog.labTemp=Temperature (deg C)
```
```suggestion
PlanetaryConditionsDialog.labTemp=Temperature (\u00B0C)
```

Windchild292 replied:
Unicode symbol, which fixes a number of possible issues (had a MekHQ bug that using the icon instead of the unicode led to display issues.

SJuliez replied:
Doesnt work on Win10 without setting the font of the label to "Dialog" (which is precisely why I hadnt used it)

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -2090,23 +2150,37 @@ PlanetaryConditionsDialog.labAtmosphere=Atmospheric Pressure
+
+PlanetaryConditionsDialog.invalid.integer=Enter an integer number.<BR>
+PlanetaryConditionsDialog.invalid.number=Enter a number.<BR>
+PlanetaryConditionsDialog.invalid.tempRange=The Temperature must be between -200 and 200 �C.<BR>
```
```suggestion
PlanetaryConditionsDialog.invalid.tempRange=The Temperature must be between -200 and 200 \u00B0C.<BR>
```

SJuliez replied:
As above

SJuliez replied:
And yes the character Ive used here doesnt work either :)

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -2090,23 +2150,37 @@ PlanetaryConditionsDialog.labAtmosphere=Atmospheric Pressure
+PlanetaryConditionsDialog.invalid.gravRange=Gravity must be between 0.1 g and 10 g.<BR>
+PlanetaryConditionsDialog.invalid.sandsLost=The effects of Blowing Sands are lost when the wind is below Moderate Gale strength.<BR>
+PlanetaryConditionsDialog.invalid.traceLightGale=In Trace Atmosphere, Light Gale is not allowed.<BR>
+PlanetaryConditionsDialog.invalid.lightSnowTemp=This weather includes the effects of a temperature of -40 �C and any higher temperature setting is invalid.<BR>
+PlanetaryConditionsDialog.invalid.modSnowTemp=This weather includes the effects of a temperature of -50 �C and any higher temperature setting is invalid.<BR>
+PlanetaryConditionsDialog.invalid.iceStormTemp=Ice Storm includes the effects of a temperature of -60 �C and any higher temperature setting is invalid.<BR>
```
```suggestion
PlanetaryConditionsDialog.invalid.lightSnowTemp=This weather includes the effects of a temperature of -40 \u00B0C and any higher temperature setting is invalid.<BR>
PlanetaryConditionsDialog.invalid.modSnowTemp=This weather includes the effects of a temperature of -50 \u00B0C and any higher temperature setting is invalid.<BR>
PlanetaryConditionsDialog.invalid.iceStormTemp=Ice Storm includes the effects of a temperature of -60 \u00B0C and any higher temperature setting is invalid.<BR>
```

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -2090,23 +2150,37 @@ PlanetaryConditionsDialog.labAtmosphere=Atmospheric Pressure
+PlanetaryConditionsDialog.invalid.iceStormTemp=Ice Storm includes the effects of a temperature of -60 �C and any higher temperature setting is invalid.<BR>
+PlanetaryConditionsDialog.invalid.windRange=The current wind strength is outside the range of shifting wind strengths.<BR>
+
+PlanetaryConditionsDialog.EnterValidTemperature=The Temperature must be between -200 and 200 �C.<BR>
```
```suggestion
PlanetaryConditionsDialog.EnterValidTemperature=The Temperature must be between -200 and 200 \u00B0C.<BR>
```

Windchild292 commented on megamek/i18n/megamek/client/messages.properties:
```diff
@@ -3539,7 +3620,7 @@ HeatEffects.tsm.mt.49=-7 MP; +7 BTH
 
 #Loading Units into Bays
-LoadingBay.baytoomany=Too many units selected.\nThe selected bay has a remaining capacity of 
+LoadingBay.baytoomany=Too many units selected.\nThe selected bay has a remaining capacity of {0}{1}. 
```
Just checking that this wasn't supposed to have a space?

SJuliez replied:
Probably...

Windchild292 commented on megamek/src/megamek/client/ratgenerator/ForceDescriptor.java:
```diff
@@ -843,6 +845,29 @@ public void loadEntities(Ruleset.ProgressListener l, double progress) {
+    
+    /** Generates a force string for exporting these units to MUL / adding to the game. */
+    private String getForceString() {
+        ArrayList<ForceDescriptor> ancestors = new ArrayList<>();
```
```suggestion
        List<ForceDescriptor> ancestors = new ArrayList<>();
```

SJuliez replied:
I'll use var instead

Windchild292 commented on megamek/src/megamek/client/Client.java:
```diff
@@ -852,6 +865,50 @@ public void sendDeploymentUnload(Entity loader, Entity loaded) {
+        
+    /** Sends a packet to the Server requesting to delete the given forces. */
+    public void sendDeleteForces(List<Force> toDelete) {
+        List<Integer> forceIds = new ArrayList<>();
+        toDelete.stream().mapToInt(f -> f.getId()).forEach(forceIds::add);
```
```suggestion
        List<Integer> forceIds = toDelete.stream().mapToInt(Force::getId).boxed().collect(Collectors.toList());
```

Windchild292 commented on megamek/src/megamek/client/Client.java:
```diff
@@ -953,7 +1015,41 @@ protected void receiveEntities(Packet c) {
+    protected void receiveForceUpdate(Packet c) {
+        Collection<Force> forces = (Collection<Force>) c.getObject(0);
+        Collection<Entity> entities = (Collection<Entity>) c.getObject(1);
+        for (Force force: forces) {
```
```suggestion
        for (Force force : forces) {
```

SJuliez replied:
See below.

Windchild292 commented on megamek/src/megamek/client/Client.java:
```diff
@@ -953,7 +1015,41 @@ protected void receiveEntities(Packet c) {
+        for (Force force: forces) {
+            game.getForces().replace(force.getId(), force);
+        }
+        for (Entity entity: entities) {
```
```suggestion
        for (Entity entity : entities) {
```

SJuliez replied:
Not my style.

Windchild292 replied:
This isn't a style bit, but code accessibility. You separate out the ":" to ensure that the symbol isn't read as part of the word and then ignored, as it makes it into a separate word.

Windchild292 commented on megamek/src/megamek/client/Client.java:
```diff
@@ -965,28 +1061,50 @@ protected void receiveEntityUpdate(Packet c) {
+     */
+    @SuppressWarnings("unchecked")
+    protected void receiveEntitiesUpdate(Packet c) {
+        Collection<Entity> entities = (Collection<Entity>) c.getObject(0);
+        for (Entity entity: entities) {
```
```suggestion
        Collection<Entity> entities = (Collection<Entity>) c.getObject(0);
        for (Entity entity : entities) {
```

Windchild292 replied:
Could also probably just have the cast in the for loop call

Windchild292 commented on megamek/src/megamek/client/Client.java:
```diff
@@ -965,28 +1061,50 @@ protected void receiveEntityUpdate(Packet c) {
+        @SuppressWarnings("unchecked")
+        List<Force> forces = (List<Force>) packet.getObject(2);
 
+        for (Force force: forces) {
```
```suggestion
        for (Force force : forces) {
```

Windchild292 commented on megamek/src/megamek/client/Client.java:
```diff
@@ -965,28 +1061,50 @@ protected void receiveEntityUpdate(Packet c) {
         for(int id: entityIds) {
             cacheImgTag(game.getEntity(id));
         }
+        for (Force force: forces) {
```
```suggestion
        for (Force force : forces) {
```

Windchild292 reviewed (changes requested):
Went through about 2/3 of this PR (including all of the data), and made comments on it.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/LobbyMekPopup.java:
```diff
@@ -0,0 +1,688 @@
+            for (int i = 1; i < 11; i++) {
+                heatMenu.add(menuItem("Heat " + i, "HEAT|" + i + eIds, enabled, listener));
+            }
+            JMenu subHeatMenu = new JMenu("More heat");
```
I didnt think so. Assigning heat should be rare. assigning 7 heat is just a menu entry. And if you really need to assign 22 heat you can do that in that last submenu. Its sure blocky but gets it done. I have to add that personally I dislike any UI item that forces me to click a tiny arrow which is why I avoid Spinners.

![image](https://user-images.githubusercontent.com/17069663/115577854-091bb580-a2c5-11eb-8f98-f6dac996a5ad.png)

NickAragua replied:
Fair enough.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/LobbyUtility.java:
```diff
@@ -0,0 +1,465 @@
+     * Returns true when the given board name does not start with one of the control strings
+     * of MapSettings signalling a random, generated or surprise board. 
+     */ 
+    @SuppressWarnings("deprecation")
```
The RANDOM entry. I replaced it with surprise.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/LobbyUtility.java:
```diff
@@ -0,0 +1,465 @@
+    /** Returns the string with some content shortened like Battle Armor -> BA */
+    static String abbreviateUnitName(String unitName) {
+        return unitName
+                .replace("(Standard)", "").replace("Battle Armor", "BA")
```
Yes. This only abbreviates the unit names for compact display. Doesnt need to be successful.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/MekForceTreeCellFormatter.java:
```diff
@@ -0,0 +1,794 @@
+        }
+        
+        // BV
+        // This takes a crazy amount of time when regiments are in the game
```
I think it won't help much. "Crazy" overstates it. I just found it crazy considering what little info-gathering this code does. With 50+ units in the game, the updates start getting noticeable. With 500 units we're speaking of around a second for an update, but changing something may trigger several updates depending on #players and other things. So things get slow. There's no single thing to change to make it faster, unfortunately. Updating stuff for each player, the icons, the text formatting, the info gathering all works together. Turning this force fluff info off won't change anything for most games and for the few ones with really many units it won't make it super fast.

SJuliez commented on megamek/src/megamek/common/Player.java:
```diff
@@ -417,6 +418,7 @@ public int getBV() {
      */
     @Override
     public int getFledBV() {
+        //TODO: I'm not sure how squadrons are treated here - see getBV()
```
Note that I changed the getBV() function you updated to a stream() a few days ago because it counted Squadrons twice. With the FledBV I just wanted to note that the same problem applied here maybe but Im not about to figure that out now.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/sorters/PlayerBVSorter.java:
```diff
@@ -0,0 +1,67 @@
+import megamek.common.Entity;
+import megamek.common.IPlayer;
+
+/** A Lobby Mek Table sorter that sorts for 1) Player 2) BV. */
+public class PlayerBVSorter implements MekTableSorter {
```
Another time...

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/sorters/C3IDSorter.java:
```diff
@@ -0,0 +1,99 @@
+    private ClientGUI clientGui;
+    
+    /** A Lobby Mek Table sorter that sorts mainly by association to C3 networks */
+    public C3IDSorter(ClientGUI cg) {
+        clientGui = cg;
+    }
+    
+    @Override
+    public String getDisplayName() {
+        return "C3 Network";
+    }
+    
+    @Override
+    public int getColumnIndex() {
+        return MekTableModel.COL_UNIT;
+    }
```
Lets not overdo it.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/sorters/ForceSorter.java:
```diff
@@ -0,0 +1,123 @@
+import megamek.common.IPlayer;
+import megamek.common.force.Force;
+
+public class ForceSorter implements MekTableSorter {
```
No longer used, deleted.

SJuliez commented on megamek/src/megamek/client/ui/swing/lobby/sorters/MekTableSorter.java:
```diff
@@ -0,0 +1,55 @@
+    int getColumnIndex();
+    
+    /**
+     * Returns true if this Sorter is currently allowed. Sorters might not be allowed e.g.
+     * when they would give away info in blind drops.
```
You do take things seriously, do you :)

SJuliez commented on megamek/src/megamek/common/Entity.java:
```diff
@@ -1482,7 +1497,21 @@ public int getBodyLocation() {
+        // object is not usually replaced. I expect entity.game to be up to date much more than owner.
+        // Unfortunately, entities freshly created may not have the game set. Therefore, fall
+        // back to the old version when game == null
+        if (game != null) {
+            return game.getPlayer(ownerId);
+        } else {
+            return owner;
+        }
+        // FORMERLY:
+        // return owner;
```
Id like this to remain non-ternary. This is kinda important, not something to collapse to a line.

SJuliez commented on megamek/src/megamek/client/ui/swing/RandomMapPanelAdvanced.java:
```diff
@@ -332,18 +332,23 @@ private void initGUI() {
         // Let each tab have it's own scroll bar.
 
         JScrollPane civilizedFeaturesPanel = setupCivilizedPanel();
+        civilizedFeaturesPanel.getVerticalScrollBar().setUnitIncrement(16);
         tabbedPane.addTab(Messages.getString("RandomMapDialog.tabCivilization"), civilizedFeaturesPanel);
 
         JScrollPane elevationPanel = setupElevationPanel();
+        elevationPanel.getVerticalScrollBar().setUnitIncrement(16);
         tabbedPane.addTab(Messages.getString("RandomMapDialog.tabElevation"), elevationPanel);
 
         JScrollPane naturalFeaturesPanel = setupNaturalFeaturesPanel();
+        naturalFeaturesPanel.getVerticalScrollBar().setUnitIncrement(16);
         tabbedPane.addTab(Messages.getString("RandomMapDialog.tabNatural"), naturalFeaturesPanel);
 
         JScrollPane effectsPanel = setupEffectsPanel();
+        effectsPanel.getVerticalScrollBar().setUnitIncrement(16);
         tabbedPane.addTab(Messages.getString("RandomMapDialog.tabEffects"), effectsPanel);
 
         JScrollPane waterPanel = setupWaterPanel();
+        waterPanel.getVerticalScrollBar().setUnitIncrement(16);
```
Um, what kinda of bug? Why, where. It is used in MM widely and I see no bugs. In any case, please give me a better solution.

SJuliez commented:
@NickAragua and @Windchild292 Went through the reviews. What I didnt comment I put in the code as suggested by the review. Nick's review changes are in two commits, Review Changes pt.1 and Nickaragua review changes; Windchild's review changes are in the latest commit.

NickAragua reviewed (approved):
Looks good from my end.

SJuliez commented:
@Windchild292 Putting this here so we don't have to search for comments. 
Accessibility means what? Someone using a screen reader to program stuff? I really don't know. If yes, has there ever been someone? Even if the : gets swallowed why would that be a problem?

And you commented on this again:
The images in this directory are temporary images used for the pilot tooltips. They can be safely deleted while not in a MegaMek game.
But this is changed in your commit. I don't see what's missing.

Any other new comments?

Windchild292 commented:
Interesting... the file did not update on my end, so ignore that comment. As for the other bit, I'll explain in a PM.
